### DISCUSSION TITLE
Code creation?

### DISCUSSION THREAD

**Discussion by adrianhensler** [assafelovic/gpt-researcher#768] [Ideas]
I wonder how this might work for code generation and what optimizations might be needed/helpful in that direction. Just as a test I ran GPT Researcher and asked for the infamous snake game. Should have asked for missile command; I'm tired of the snake game - on a short run and got this workable result:
```

import pygame
import time
import random

# Initialize Pygame
pygame.init()

# Define colors
white = (255, 255, 255)
yellow = (255, 255, 102)
black = (0, 0, 0)
red = (213, 50, 80)
green = (0, 255, 0)
blue = (50, 153, 213)

# Set display dimensions
dis_width = 800
dis_height = 600

# Create game window
dis = pygame.display.set_mode((dis_width, dis_height))
pygame.display.set_caption('Snake Game by Edureka')

# Set clock
clock = pygame.time.Clock()

# Set snake block size and speed
snake_block = 10
snake_speed = 15

# Set fonts
font_style = pygame.font.SysFont("bahnschrift", 25)
score_font = pygame.font.SysFont("comicsansms", 35)

def our_snake(snake_block, snake_List):
    for x in snake_List:
        pygame.draw.rect(dis, black, [x[0], x[1], snake_block, snake_block])

def message(msg, color):
    mesg = font_style.render(msg, True, color)
    dis.blit(mesg, [dis_width / 6, dis_height / 3])

def gameLoop():
    game_over = False
    game_close = False

    x1 = dis_width / 2
    y1 = dis_height / 2

    x1_change = 0
    y1_change = 0

    snake_List = []
    Length_of_snake = 1

    foodx = round(random.randrange(0, dis_width - snake_block) / 10.0) * 10.0
    foody = round(random.randrange(0, dis_height - snake_block) / 10.0) * 10.0

    while not game_over:

        while game_close == True:
            dis.fill(blue)
            message("You Lost! Press Q-Quit or C-Play Again", red)
            pygame.display.update()

            for event in pygame.event.get():
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_q:
                        game_over = True
                        game_close = False
                    if event.key == pygame.K_c:
                        gameLoop()

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                game_over = True
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT:
                    x1_change = -snake_block
                    y1_change = 0
                elif event.key == pygame.K_RIGHT:
                    x1_change = snake_block
                    y1_change = 0
                elif event.key == pygame.K_UP:
                    y1_change = -snake_block
                    x1_change = 0
                elif event.key == pygame.K_DOWN:
                    y1_change = snake_block
                    x1_change = 0

        if x1 >= dis_width or x1 < 0 or y1 >= dis_height or y1 < 0:
            game_close = True
        x1 += x1_change
        y1 += y1_change
        dis.fill(blue)
        pygame.draw.rect(dis, green, [foodx, foody, snake_block, snake_block])
        snake_Head = []
        snake_Head.append(x1)
        snake_Head.append(y1)
        snake_List.append(snake_Head)
        if len(snake_List) > Length_of_snake:
            del snake_List[0]

        for x in snake_List[:-1]:
            if x == snake_Head:
                game_close = True

        our_snake(snake_block, snake_List)
        pygame.display.update()

        if x1 == foodx and y1 == foody:
            foodx = round(random.randrange(0, dis_width - snake_block) / 10.0) * 10.0
            foody = round(random.randrange(0, dis_height - snake_block) / 10.0) * 10.0
            Length_of_snake += 1

        clock.tick(snake_speed)

    pygame.quit()
    quit()

gameLoop()
```

**Comment by adrianhensler** [comment]
Likely just the wrong tool for the job; hammer for a screw.

**Reply by danieldekay** [reply]
agreed - here's one example that does what you need: https://github.com/geekan/MetaGPT

PS: it has a lof of similarity with GPT-researcher's multiple-agents feature, but the prompts are different.

**Reply by ElishaKay** [reply]
Oh. Now you're getting me started!

This tool is brilliant for this use case:

https://app.greptile.com (definitely worth the $20 monthly, in my humble opinion)

There's a couple other options out there:

[OpenDevin](https://github.com/OpenDevin/OpenDevin) (tried it once but I like Greptile much better)

The ability of the AI Agent to test the code recommendations in Docker Containers will be huge - I need to look deeper into it, so let me know if you find something good

**Comment by ElishaKay** [comment]
@adrianhensler 

Agreed. We now have a hybrid report source which is able to take both a Document Path and Search Engine as input 

Meaning, it could start off looking through a code project & then search the web for recommendations for that same codebase

Would be great to create a new multi -Agents flow for this type of use case: coding copilot.

It should have the ability to spin up docker containers and test the code it recommends

**Reply by danieldekay** [reply]
don't know if I agree. While I love the idea, I think it diverts resources away from the core use case... no?

**Reply by ElishaKay** [reply]
Think about the functions & data flow that make GPTR work: 

 - parsing chunks and indexing them 
 - searching through those chunks based on user queries
 - searching the web (ie giving the agents the ability to leverage existing API's) and putting together a response based on the user queries
 - looping in the user as the Agent progresses and asking for guidance 

Feels like the same abstractions under the hood which could be applied to a wide variety of use cases.

My feeling is that we're just at the very beginning of finding the right tools, libraries and design patterns to meet these use cases.

That's the beauty of open source - worse case, your PR will verge off master 🍣

Follow your imagination 🥠

**Reply by ElishaKay** [reply]
Also, put [this as background music](https://open.spotify.com/track/6k41naacsmD4L3i9AzCFom?si=WGpxotnBT7qcMbLzIauXWA) while reading the above:

🤣

**Reply by danieldekay** [reply]
Hahaha.

Yes, I agree that the core process is the same.

If GPTR had an easy way to specify a team of agents, and their prompts. That's essentially what custom_reports would be. It's just the question of how someone uses GPTR. Through the frontend or as package.

**Reply by danieldekay** [reply]
But let's say: adding code execution in docker container sandboxes, etc -- is something that others are building with a core focus only on code generation.

**Reply by ElishaKay** [reply]
@danieldekay you're making some good points - I'll need to meditate upon them

 @assafelovic says:

`The long term vision is being the #1 research agent for developers and multi agent frameworks`

**Reply by pedrorocha-net** [reply]
Great discussion and idea guys!

IMHO, from the perspective of product management, though, would maybe make sense to create a separate project that uses GPTR as a package, but not turning a Research Tool into a Coding Tool. Completely different audiences and use cases, so bringing that might bring a lot of confusion to the documentation, landing page, any communication piece, as the target audiences will be very different.

**Comment by a-yuksel** [comment]
I am using [Autogen](https://microsoft.github.io/autogen/)  where you can easily include agents to write code as well as **_EXECUTE_** the code (in your local environment or in a docker container). It is a competitor to Langchain's agent framework. I would strongly suggest it for code execution

**Reply by assafelovic** [reply]
Wdyt of opening a PR for it? Would love to see an implementation of GPTR with autogen!

**Reply by a-yuksel** [reply]
I am working on using GPTResearcher as an agent in Autogen framework. Happy to share it when completed

**Comment by ElishaKay** [comment]
Welcome @pedrorocha-net,

it's a fair point. 

Our core focus for the upcoming sprint is a demo app which will open up the platform to Non-Technical users.

That being said: relative to the awesome power of the LLM, the existing AI coding tools:

-  leave a lot to be desired functionality wise
-  can be quite expensive
-  are difficult to customize 
-  & could introduce security concerns

Think about how much mental capacity is wasted on a weekly basis by developers looking up error stack traces & researching how to implement new features in complex codebases - especially in a microservices environment where 1 feature is often spread across 3 or more services.

Within our own GPTR community we're considering a discord bot to help with our documentation & FAQ efforts.

I opened up a PR for this leveraging some of Assaf's design patterns & helper functions from the multi_agents flow:

https://github.com/assafelovic/gpt-researcher/pull/819

If it can add value to our community, hopefully we'll get an ROI on it at least within the GPTR tribe of developers building the next-gen AI framework... with AI 

cc: @a-yuksel - feel free to ping me for a collab on this one
